### PR TITLE
Update to ungoogled-chromium 123.0.6312.58

### PR DESCRIPTION
**Pull request by teeminus**
Builds and runs fine:

image

image

However, there are quite some changes in this release. As rust gets integrated more and more into the chromium source code, I implemented using the rust-toolchain in our build process. To give you guys more time to review the changes (or try them yourselfs) I created this PR prior the the official release of ungoogled-chromium 123.0.6312.46-1. In order to test these changes you need to update the submodule manually.

(As always,) Google ships its own precompiled version of the rust-toolchain in a dedicated archive file thats gets downloaded during the build process. I downloaded and extracted that archive and tried to figure out how to rebuild/replace it using the offical rust releases.

The folder structure of the Google rust archive is the following:
<details>
<summary>find * -maxdepth 1 -type d</summary>
```
bin
etc
etc/bash_completion.d
lib
lib/rustlib
share
share/doc
share/man
share/zsh
```
</details>

The folder structure of the offical rust archive is the following:
<details>
<summary>find * -maxdepth 1 -type d</summary>
```
$ find * -maxdepth 1 -type d
cargo
cargo/bin
cargo/etc
cargo/share
clippy-preview
clippy-preview/bin
clippy-preview/share
llvm-tools-preview
llvm-tools-preview/lib
rls-preview
rls-preview/bin
rls-preview/share
rust-analysis-x86_64-pc-windows-msvc
rust-analysis-x86_64-pc-windows-msvc/lib
rust-analyzer-preview
rust-analyzer-preview/bin
rust-analyzer-preview/share
rust-demangler-preview
rust-demangler-preview/bin
rust-demangler-preview/share
rust-docs
rust-docs/share
rust-docs-json-preview
rust-docs-json-preview/share
rust-std-x86_64-pc-windows-msvc
rust-std-x86_64-pc-windows-msvc/lib
rustc
rustc/bin
rustc/lib
rustc/libexec
rustc/share
rustfmt-preview
rustfmt-preview/bin
rustfmt-preview/share
```
</details>

It turned out, that the Google archive can be replicated (mostly) by copying/merging all [`bin`, `lib`] folders from the toplevel folders of the official rust archive. (I already stripped down the number of folders to merge.) I automated that by adding a step to the `build.py` script to create the `build/src/third_party/rust-toolchain` folder. This process depends on the target architecture, thats why we need to download both the x64 and the x86 version of the offical rust archive.

Finding the correct rust version for the builds process is an other topic. Google uses a nighly build (currently 1.78) of rust, the latest offical release is 1.76. We could probably use the 1.76 rust release, but that would required patching a lot more files (to patch lib dependencies and supported compiler flags). I think we won't be able to get the exact same rust version Google is using in their recompiled toolchain, but we can find a somewhat close official nightly release. The required rust-version is hard coded in the `build/src/tools/rust/update_rust.py` file in the `RUST_REVISION` constant. Using that commit hash we can find the commit date and use that date to download the official rust nightly build. We could of course always use the latest nightly release but that could lead to different behaviour in the end. I added a description of how the check/update the rust version to the `README.MD`.

Other noteable changes:
- Update to clang 18.1.1
- The patch file `ungoogled-chromium/windows/windows-fix-missing-include-es_parser_adts-cc.patch` is no longer required

### CODE REVIEW

**Comment by PF4Public** [comment]
> In order to test these changes you need to update the submodule manually.

Would it make sence to rethink this structure (the submodule) for the future?

**Comment by teeminus** [comment]
> Would it make sense to rethink this structure (the submodule) for the future?

I think using the submodule is fine, it's just a minor inconvenience when preparing an update that has not yet been merged in the core repo.

> I'd recommend using glob to traverse rust dirs.

Thanks, I integrated your proposed changes.


Currently, I am struggling with building the X86 release. I suspect the issue beeing caused by a change of the `rust_abi_target` during the build process but I am not sure about that. Unfortunately it takes by build machine close to one day to compile a release so it's quite time consuming and frustrating to test fixes for the issue.

**Comment by teeminus** [comment]
There are not patch/list updates required for 123.0.6312.58, so this PR also works with the update PR from the core repo.

File: patches/ungoogled-chromium/windows/windows-fix-building-with-rust.patch [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+   # <home dir>/.rustup/toolchains/nightly-<something>-<something>
+-  rust_sysroot_absolute = ""
++  rust_sysroot_absolute = "../../third_party/rust-toolchain"
+ 
+   # If you're using a Rust toolchain as specified by rust_sysroot_absolute,
+   # set this to the output of `rustc -V`. Changing this string will cause all
+   # Rust targets to be rebuilt, which allows you to update your toolchain and
+   # not break incremental builds.
+-  rustc_version = ""
++  rustc_version = "rustc 1.76.0 (07dca489a 2024-02-04)"
```
Comment (Nifury) [inline patches/ungoogled-chromium/windows/windows-fix-building-with-rust.patch:26]: Shouldn't it be "rustc 1.78.0-nightly (a84bb95a1 2024-02-13)"?
Doesn't seem to matter though.

**Comment by Nifury** [comment]
> Currently, I am struggling with building the X86 release. I suspect the issue beeing caused by a change of the rust_abi_target during the build process

It might be a known issue on x86: https://issues.chromium.org/issues/40275898#comment11
Files in `win_clang_x86_for_rust_host_build_tools` are all 64-bit.

**Comment by Nifury** [comment]
Turns out the rust toolchain version needs to match the host cpu. 
`rust_host_toolchain_arch = host_cpu`
I guess that means we don't need to download x86 rust toolchain anymore. 
image

File: patches/ungoogled-chromium/windows/windows-fix-building-with-rust.patch (hunk 1 above)
Comment (teeminus) [reply patches/ungoogled-chromium/windows/windows-fix-building-with-rust.patch]: You are right, forgot to change that after I updated to the nighly build.

**Comment by teeminus** [comment]
> Turns out the rust toolchain version needs to match the host cpu. `rust_host_toolchain_arch = host_cpu` I guess that means we don't need to download x86 rust toolchain anymore.

I tried compiling with just the x64 rust toolchain but it failed because the x86 libs were missing. But compiling succeeded with the x64 bin folder and both x64 and x86 libs copied to the toolchain folder.

**Comment by teeminus** [comment]
As compiling for x86 succeeded now, I think this PR is ready. Do we want to recompile one more time to verify the results?

**Review by Nifury — APPROVED** [review]
I believe it's ready 😀

**Pull request merged**